Java Psych fails in ugly ways when SnakeYAML version is wrong

Originally reported via https://github.com/jruby/jruby/issues/6023

When the version of SnakeYAML loaded does not match the one required by JRuby's Psych extension, it produces a very arcane Java error:

```
% jruby -S bundle install
Ignoring bcrypt-ruby-3.0.0-java because its extensions are not built. Try: gem pristine bcrypt-ruby --version 3.0.0
Ignoring bindex-0.5.0 because its extensions are not built. Try: gem pristine bindex --version 0.5.0
Ignoring sassc-1.11.4 because its extensions are not built. Try: gem pristine sassc --version 1.11.4
/home/enebo/work/jruby/lib/ruby/stdlib/psych.rb:7: warning: already initialized constant SNAKEYAML_VERSION
/home/enebo/work/jruby/lib/ruby/stdlib/psych.rb:7: warning: already initialized constant ANY
/home/enebo/work/jruby/lib/ruby/stdlib/psych.rb:7: warning: already initialized constant UTF8
/home/enebo/work/jruby/lib/ruby/stdlib/psych.rb:7: warning: already initialized constant UTF16LE
/home/enebo/work/jruby/lib/ruby/stdlib/psych.rb:7: warning: already initialized constant UTF16BE
org/jruby/RubyKernel.java:984: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
LoadError: load error: psych -- java.lang.NoClassDefFoundError: Could not initialize class org.jruby.ext.psych.PsychEmitter
LoadError: load error: psych -- java.lang.NoClassDefFoundError: Could not initialize class org.jruby.ext.psych.PsychEmitter
```

This is caused by an API mismatch; SnakeYAML changed its "Mark" API in an incompatible way. If the old SnakeYAML loads first, loading of the Psych extension continues but eventually fails when it tries to call this API.

We should try to modify the JRuby version of Psych to fail earlier when SnakeYAML is not the expected version. I'll look into this a bit today.

This part of the warning is confusing and I think not applicable for JRuby:
```
org/jruby/RubyKernel.java:984: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
```

--- a/ext/java/org/jruby/ext/psych/PsychLibrary.java
+++ b/ext/java/org/jruby/ext/psych/PsychLibrary.java
@@ -44,8 +44,11 @@ import org.jruby.runtime.ThreadContext;
 import org.jruby.runtime.Visibility;
 import org.jruby.runtime.builtin.IRubyObject;
 import org.jruby.runtime.load.Library;
+import org.yaml.snakeyaml.error.Mark;
 
 public class PsychLibrary implements Library {
+    private static final String DUMMY_VERSION = "0.0";
+
     public void load(final Ruby runtime, boolean wrap) {
         RubyModule psych = runtime.defineModule("Psych");
 
@@ -57,10 +60,31 @@ public class PsychLibrary implements Library {
         catch( IOException e ) {
             // ignored
         }
-        String snakeyamlVersion = props.getProperty("version", "0.0");
+        String snakeyamlVersion = props.getProperty("version", DUMMY_VERSION);
 
         if (snakeyamlVersion.endsWith("-SNAPSHOT")) {
             snakeyamlVersion = snakeyamlVersion.substring(0, snakeyamlVersion.length() - "-SNAPSHOT".length());
+        }
+
+        // Try to determine if we have a new enough SnakeYAML.
+        // Versions before 1.21 removed a Mark constructor that JRuby uses.
+        // See https://github.com/bundler/bundler/issues/6878
+        if (snakeyamlVersion.equals(DUMMY_VERSION)) {
+            try {
+                // Use reflection to try to confirm we have a new enough version
+                Mark.class.getConstructor(String.class, int.class, int.class, int.class, int[].class, int.class);
+            } catch (NoSuchMethodException nsme) {
+                throw runtime.newLoadError("bad SnakeYAML version, required 1.21 or higher; check your CLASSPATH for a conflicting jar");
+            }
+        } else {
+            // Parse version string to check for 1.21+
+            String[] majorMinor = snakeyamlVersion.split("\\.");
+
+            if (majorMinor.length < 2 || Integer.parseInt(majorMinor[0]) < 1 || Integer.parseInt(majorMinor[1]) < 21) {
+                throw runtime.newLoadError(
+                        "bad SnakeYAML version " + snakeyamlVersion +
+                                ", required 1.21 or higher; check your CLASSPATH for a conflicting jar");
+            }
         }
 
         RubyString version = runtime.newString(snakeyamlVersion + ".0");

